StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/CustomerSheet/CustomerSheetDataSource.swift

```diff
@@ -0,0 +1,164 @@
+        Task {
+            do {
+                async let paymentMethodsResult = try customerAdapter.fetchPaymentMethods()
+                async let selectedPaymentMethodResult = try customerAdapter.fetchSelectedPaymentOption()
+                async let elementsSessionResult = try self.configuration.apiClient.retrieveElementsSessionForCustomerSheet(paymentMethodTypes: customerAdapter.paymentMethodTypes, customerSessionClientSecret: nil)
+
+                // Ensure local specs are loaded prior to the ones from elementSession
+                await loadFormSpecs()
+
+                let (paymentMethods, selectedPaymentMethod, elementSession) = try await (paymentMethodsResult, selectedPaymentMethodResult, elementsSessionResult)
+
+                // Override with specs from elementSession
+                _ = FormSpecProvider.shared.loadFrom(elementSession.paymentMethodSpecs as Any)
+
+                completion(.success((paymentMethods, selectedPaymentMethod, elementSession)))
+            } catch {
+                completion(.failure(error))
+            }
+        }
+    }
+
+    func loadFormSpecs() async {
+        await withCheckedContinuation { continuation in
+            Task {
+                FormSpecProvider.shared.load { _ in
+                    continuation.resume()
+                }
+            }
+        }
+    }
+}
+
+extension CustomerSheetDataSource {
+    func merchantSupportedPaymentMethodTypes(elementsSession: STPElementsSession) -> [STPPaymentMethodType] {
+        switch dataSource {
+        case .customerSession:
+            return elementsSession.orderedPaymentMethodTypes
+        case .customerAdapter(let customerAdapter):
+            return customerAdapter.canCreateSetupIntents ? elementsSession.orderedPaymentMethodTypes : [.card]
+        }
+    }
+
+    var canCreateSetupIntents: Bool {
+        switch dataSource {
+        case .customerAdapter(let customerAdapter):
+            return customerAdapter.canCreateSetupIntents
+        case .customerSession:
+            return true
+        }
+    }
+
+    func attachPaymentMethod(_ paymentMethodId: String) async throws {
+        switch dataSource {
+        case .customerAdapter(let customerAdapter):
+            try await customerAdapter.attachPaymentMethod(paymentMethodId)
+        case .customerSession:
+            assertionFailure("Attach payment methods are not supported with CustomerSessions")
+        }
+    }
+
+    func fetchSavedPaymentMethods() async throws -> [STPPaymentMethod]? {
+        switch dataSource {
+        case .customerAdapter(let customerAdapter):
+            return try await customerAdapter.fetchPaymentMethods()
+        case .customerSession(let customerSessionAdapter):
+            let elementsSessionResult = try await customerSessionAdapter.elementsSession()
+            return elementsSessionResult.customer?.paymentMethods ?? []
```

Comment:
Should this return nil or an empty array?

Reply:
Oof, great catch.  In the event that we fail to get payment methods, we need to return nil, which corresponds to dismissing the sheet and force users to effectively re-open the sheet.